File: revision/square_of_sorted_array.py
from typing import List

def sortedSquares(nums:List[int])->List[int]:
    #O(N) Two pointer approach
    ans = [None for _ in nums]
    n, l, r = len(nums) -1 , 0, len(nums) - 1
    while n >= 0:
        if abs(nums[l]) > abs(nums[r]):
            ans[n] = nums[l] ** 2
            l+=1
            n-=1
        else:
            ans[n] = nums[r] ** 2
            r-=1
            n-=1
    return ans

nums = list(map(int, input().split()))
print(sortedSquares(nums))

File: revision/square_ofsorted_array.cpp
#include<iostream>
#include<vector>
#include<math.h>
#include<algorithm>
using namespace std;
void input(vector<int>&nums){
   for(auto &i : nums){
       cin >> i;
   }
}
void display(vector<int>nums){
    for(auto const i : nums){
        cout << i << ' ' ;
    }
    cout << endl;
}

vector<int> sortedSquares(vector<int>&nums){
    //Nlog(N) approach using sort:-
    // for(int i = 0; i < nums.size() ; i++){
    //     nums[i] = nums[i] * nums[i];        
    // }
    // sort(nums.begin(), nums.end());
    // return nums;
    
    //O(N) approach and O(N)SC using 2 pointers:
    vector<int>ans(nums.begin(), nums.end());
    int n = nums.size() - 1 , l = 0, r = n;
    while(n >= 0){
        if ((nums[l] * nums[l]) > (nums[r] * nums[r])){
            ans[n] = nums[l] * nums[l];
            l++, n--;
        } 
        else{
            ans[n] = nums[r] * nums[r];
            r--, n--;    
        }
    }
    return ans;
}
int main()
{
    int n;
    cin >> n;
    vector<int>nums(n);
    input(nums);
    display(nums);
    display(sortedSquares(nums));
    return 0;
}